README
Library for autodection charset of russian text

ChangeLog
0.1.7:
    * Don't use strnlen

0.1.6:
    * Function name is changed to conform LibRCC naming conventions
    * Other minor changes

0.1.5:
    * Few speed optimizations
    
0.1.4:
    * Fixed 'const' declarations 

csa19:
    * CP866 support

csa18: 
    * Fixed error in binary search algorithm 
    * UTF-8 support integrated
    * Improoved algorithm and extended statistics tables are implemented
    * Support for upercase phrases
    * The probability of right answers was rised.
	- Only 0.3% of mistakes distinguishing between KOI8-R and CP1251
csa8: 
    * Changed to support limited size strings 

csa:
    * Original idea was stealed from GPPL's (http://gppl.terminal.ru) 
    autorecode-xmms1.2.7 patch
    * Can autodetect CP1251 and KOI8-R lowercase phrases


statgen/test.c
#include <stdio.h>
#include "librcd.tmp.c"

main(int argc, char *argv[]) {
    FILE *f;
    int len,st;
    char word[256],phrase[8192];
    unsigned long a[4]={0,0,0,0};
    int i,max,mw;


    if ((argc!=2)&&(argc!=3)) {
	printf("Usage: %s <file name> [<max words>]\n",argv[0]);
	exit(0);
    }

    if (argc==3) mw=atoi(argv[2]);
    else mw=1;
    
    f=fopen(argv[1],"r");
    if (!f) {
	printf("Failed to open specified file. Check permissions!\n");
	exit(1);
    }
    
    while(!feof(f)) {
	strcpy(phrase,"");
	for (i=0;i<mw;i++) {
	    if (i) strcat(phrase," ");
	    fscanf(f,"%s",&word);
	    for(st=0;word[st]=='"'||word[st]=='\''||word[st]=='(';st++);
	    if (strlen(word)<1) continue;
	    for(len=strlen(word)-1;word[len]==','||word[len]=='.'||word[len]=='!'||word[len]=='?'||word[len]==';'||word[len]=='-'||word[len]==':'||word[len]=='"'||word[len]=='\''||word[len]==')';len--);
	    if (strlen(word)<1) continue;
	    else word[len+1]=0;
	    strcat(phrase,word+st);
	}
	if (strlen(phrase)<5) continue;

	a[rcdGetRussianCharset(phrase,0)]++;
    }

    printf("Win: %lu, Koi: %lu, Alt: %lu, UTF: %lu\n",a[0],a[1],a[3],a[2]);
    fclose(f);
    if (a[0]>a[1]) {
	if (a[0]>a[2]) max=0;
	else max=2;
    } else {
	if (a[1]>a[2]) max=1;
	else max=2;
    }
    if (a[3]>max) max=3;

    f=fopen(argv[1],"r");
    while(!feof(f)) {
	strcpy(phrase,"");
	for (i=0;i<mw;i++) {
	    if (i) strcat(phrase," ");
	    fscanf(f,"%s",&word);
	    for(st=0;word[st]=='"'||word[st]=='\''||word[st]=='(';st++);
	    if (strlen(word)<1) continue;
	    for(len=strlen(word)-1;word[len]==','||word[len]=='.'||word[len]=='!'||word[len]=='?'||word[len]==';'||word[len]=='-'||word[len]==':'||word[len]=='"'||word[len]=='\''||word[len]==')';len--);
	    if (strlen(word)<1) continue;
	    else word[len+1]=0;
	    strcat(phrase,word+st);
	}
	if (strlen(phrase)<5) continue;

	i=rcdGetRussianCharset(phrase,0);
	if (i!=max) {
	    if (i==0) printf("Win: %s\n",phrase);
	    else if (i==1) printf("Koi: %s\n",phrase);
	    else if (i==2) printf("UTF: %s\n",phrase);
	    else if (i==3) printf("ALT: %s\n",phrase);
	}
    }
    fclose(f);
}


src/librcd.c
#include <stdio.h>

#define _LIBRCD_C
#include "librcd.h"

#define NF_VALUE -2
#define max(a,b) ((a>b)?a:b)
#define min(a,b) ((a<b)?a:b)
#define bit(i) (1<<i)

typedef struct lng_stat2 {
  unsigned char a;
  unsigned char b;
  double rate;
  double srate;
  double erate;
} lng_stat2;

#include "russian_table.h"


static int end_symbol(char ch) {
    if (ch=='\r'||ch=='\n'||ch==0||ch==' '||ch=='\t'||ch==','||ch=='.'||ch=='!'||ch=='?'||ch==';'||ch=='-'||ch==':'||ch=='"'||ch=='\''||ch==')') return 1;
    return 0;
}

static int start_symbol(char ch) {
    if ((ch=='\t')||ch=='\r'||ch=='\n'||(ch==' ')||(ch=='(')||(ch=='"')||(ch=='\'')) return 1;
    return 0;
}    

typedef const struct lng_stat2 *lng_stat2_ptr;

static void bfind(const unsigned char *a, lng_stat2_ptr *w, lng_stat2_ptr *k, lng_stat2_ptr *al) {
  const struct lng_stat2 *winptr, *koiptr,*altptr;
  int ki,wi,ai,d,ws=0,ks=0,as=0;
    d=npow2>>1;
    wi=d;
    ki=d;
    ai=d;
    winptr=0;
    koiptr=0;
    altptr=0;
    do{
      d>>=1;
    
      if(!ws){
       if (wi>indexes2) wi-=d;
       else {
        winptr=enc_win+wi-1;
        if(a[0]==winptr->a){
          if(a[1]==winptr->b){
        	ws=1;
          }else if(a[1]<winptr->b){
            wi-=d;
          }else{ //b>win[wi].b
            wi+=d;
          }
        }else if(a[0]<winptr->a){
          wi-=d;
        }else{ //a>win[wi].a
          wi+=d;
        }
       }
      }
      if(!ks){
       if (ki>indexes2) ki-=d;
       else {
        koiptr=enc_koi+ki-1;
        if(a[0]==koiptr->a){
          if(a[1]==koiptr->b){
        	ks=1;
          }else if(a[1]<koiptr->b){
            ki-=d;
          }else{ //b>win[wi].b
            ki+=d;
          }
        }else if(a[0]<koiptr->a){
          ki-=d;
        }else{ //a>win[wi].a
          ki+=d;
        }
       }
      }
      if(!as){
       if (ai>indexes2) ai-=d;
       else {
        altptr=enc_alt+ai-1;
        if(a[0]==altptr->a){
          if(a[1]==altptr->b){
        	as=1;
          }else if(a[1]<altptr->b){
            ai-=d;
          }else{ //b>win[wi].b
            ai+=d;
          }
        }else if(a[0]<altptr->a){
          ai-=d;
        }else{ //a>win[wi].a
          ai+=d;
        }
       }
      }
    }while(d);
    if (ws) *w=winptr;
    else *w=NULL;
    if (ks) *k=koiptr;
    else *k=NULL;
    if (as) *al=altptr;
    else *al=NULL;
}

static double calculate(double s, double m, double e) {
    return s+m+e;
}

static int is_win_charset2(const unsigned char *txt, int len){
  const struct lng_stat2 *winptr, *koiptr,*altptr;
  double winstep,koistep,altstep,winestep,koiestep,altestep,winsstep,koisstep,altsstep;
  double winstat=0,koistat=0,altstat=0,winestat=0,koiestat=0,altestat=0,winsstat=0,koisstat=0,altsstat=0;
  long j;
  
#ifdef _AUTO_DEBUG
  fprintf(stderr,"Word: %s\n",txt);
#endif
  for(j=0;j<len-1;j++){
    //skip bottom half of table
    if(txt[j]<128 || txt[j+1]<128) continue;
#ifdef _AUTO_DEBUG
    fprintf(stderr,"Pair: %c%c",txt[j],txt[j+1]);
#endif
    bfind(txt+j,&winptr,&koiptr,&altptr);

    if ((j==0)||(start_symbol(txt[j-1]))) {
	if (winptr) winsstep=winptr->srate;
	else winsstep=NF_VALUE;
	if (koiptr) koisstep=koiptr->srate;
	else koisstep=NF_VALUE;
	if (altptr) altsstep=altptr->srate;
	else altsstep=NF_VALUE;
	winestep=0;
	koiestep=0;
	altestep=0;
	winstep=0;
	koistep=0;
	altstep=0;
#ifdef _AUTO_DEBUG
	fprintf(stderr,", Win %lf, Koi %lf, Alt: %lf\n",winsstep,koisstep,altsstep);
#endif
    } else if ((j==len-2)||(end_symbol(txt[j+2]))) {
	if (winptr) winestep=winptr->erate;
	else winestep=NF_VALUE;
	if (koiptr) koiestep=koiptr->erate;
	else koiestep=NF_VALUE;
	if (altptr) altestep=altptr->erate;
	else altestep=NF_VALUE;
	winsstep=0;
	koisstep=0;
	altsstep=0;
	winstep=0;
	koistep=0;
	altstep=0;
#ifdef _AUTO_DEBUG
	fprintf(stderr,", Win %lf, Koi %lf, Alt %lf\n",winestep,koiestep,altestep);
#endif
    } else {
	if (winptr) winstep=winptr->rate;
	else winstep=NF_VALUE;
	if (koiptr) koistep=koiptr->rate;
	else koistep=NF_VALUE;
	if (altptr) altstep=altptr->rate;
	else altstep=NF_VALUE;
	winsstep=0;
	winestep=0;
	koisstep=0;
	koiestep=0;
	altsstep=0;
	altestep=0;
#ifdef _AUTO_DEBUG
	fprintf(stderr,", Win %lf, Koi %lf, Alt %lf\n",winstep,koistep,altstep);
#endif
    }
    
    winstat+=winstep;
    koistat+=koistep;
    altstat+=altstep;
    winsstat+=winsstep;
    koisstat+=koisstep;
    altsstat+=altsstep;
    winestat+=winestep;
    koiestat+=koiestep;
    altestat+=altestep;
  }
  
#ifdef _AUTO_DEBUG
  fprintf(stderr,"Start.  Win: %lf, Koi: %lf, Alt: %lf\n",winsstat,koisstat,altsstat);
  fprintf(stderr,"Middle. Win: %lf, Koi: %lf, Alt: %lf\n",winstat,koistat,altstat);
  fprintf(stderr,"End.    Win: %lf, Koi: %lf, Alt: %lf\n",winestat,koiestat,altestat);
  fprintf(stderr,"Final.  Win: %lf, Koi: %lf, Alt: %lf\n",calculate(winsstat,winstat,winestat),calculate(koisstat,koistat,koiestat),calculate(altsstat,altstat,altestat));
#endif
  if ((calculate(altsstat,altstat,altestat)>calculate(koisstat,koistat,koiestat))&&(calculate(altsstat,altstat,altestat)>calculate(winsstat,winstat,winestat))) return 3;
  if (calculate(koisstat,koistat,koiestat)>calculate(winsstat,winstat,winestat)) return 1;
  return 0;
}


static int check_utf8(const unsigned char *buf, int len) {
    long i,j;
    int bytes=0,rflag=0;
    unsigned char tmp;
    int res=0;

    for (i=0;i<len;i++) {
	if (buf[i]<128) continue;
	
	if (bytes>0) {
	    if ((buf[i]&0xC0)==0x80) {
		if (rflag) {
		    tmp=buf[i]&0x3F;
		    // Russian is 0x410-0x44F
		    if ((rflag==1)&&(tmp>=0x10)) res++;
		    else if ((rflag==2)&&(tmp<=0x0F)) res++;
		}
		bytes--;
	    } else {
		res--;
		bytes=1-bytes;
		rflag=0;
	    }
	} else {
	    for (j=6;j>=0;j--)
		if ((buf[i]&bit(j))==0) break;
	    
	    if ((j==0)||(j==6)) {
		if ((j==6)&&(bytes<0)) bytes++;
		else res--;
		continue;
	    }
	    bytes=6-j;
	    if (bytes==1) {
		// Cyrrilic D0-D3, Russian - D0-D1
		if (buf[i]==0xD0) rflag=1;
		else if (buf[i]==0xD1) rflag=2;
	    }
	}
	    
	if ((buf[i]==0xD0)||(buf[i]==0xD1)) {
	    if (i+1==len) break;
	    	    
	}
    }
    return res;
}



rcd_russian_charset rcdGetRussianCharset(const char *buf,int len) {
    long l;

    l = len?len:strlen(buf);
    if (check_utf8(buf,l)>1) return RUSSIAN_CHARSET_UTF8;
    return is_win_charset2(buf,l);
}

/* Compatibility */
rcd_russian_charset get_russian_charset(const char *buf,int len) {
    return rcdGetRussianCharset(buf, len);
}


src/librcd.h
#ifndef _LIBRCD_H
#define _LIBRCD_H

#ifdef __cplusplus
extern "C" {
#endif

enum rcd_russian_charset_t {
    RUSSIAN_CHARSET_WIN = 0,
    RUSSIAN_CHARSET_KOI,
    RUSSIAN_CHARSET_UTF8,
    RUSSIAN_CHARSET_ALT
};
typedef enum rcd_russian_charset_t rcd_russian_charset;


/*
rcdGetRussianCharset
    Detects encoding of russian text passed in buf variable. Support 
    UTF-8, CP1251, CP866 and KOI8-R encoding.
    
    buf  	String with russian texts
    len  	Number of bytes to use from buf. If zero is passed determined 
		by strlen.
    
    Returns:
	0 - CP1251
	1 - KOI8-R
	2 - UTF8
	3 - CP866
*/

rcd_russian_charset rcdGetRussianCharset(const char *buf, int len);

#ifdef __cplusplus
}
#endif

/* Backward compatibility */
#ifndef _LIBRCD_C
# define russian_charsets rcd_russian_charset
# define get_russian_charset rcdGetRussianCharset
#endif /* ! _LIBRCD_C */

#endif /* _LIBRCD_H */


statgen/upper.c
#include <stdio.h>
#include <unistd.h>
#include <ctype.h>
#include <locale.h>
#include <sys/types.h>
#include <sys/stat.h>


main(int argc, char *argv[]) {
    FILE *f;
    struct stat st;
    unsigned char *text;
    char locale[32];
    int i;


    if (argc!=3) {
	printf("Usage: %s <file name> <encoding>\n",argv[0]);
	exit(0);
    }

    if (strlen(argv[2])>12) {
	printf("Invalid encoding(%s) specified!\n",argv[2]);
	exit(1);
    }
    
    if ((!strcasecmp(argv[2],"koi"))||(!strcasecmp(argv[2],"koi8"))||(!strcasecmp(argv[2],"koi-8"))||(!strcasecmp(argv[2],"koi8-r")))
    	sprintf(locale,"ru_RU.%s","KOI8-R");
    else if ((!strcasecmp(argv[2],"win"))||(!strcasecmp(argv[2],"cp1251"))||(!strcasecmp(argv[2],"cp-1251"))||(!strcasecmp(argv[2],"win1251"))||(!strcasecmp(argv[2],"win-1251")))
	sprintf(locale,"ru_RU.%s","CP1251");
    else
	sprintf(locale,"ru_RU.%s",argv[2]);
    if (!setlocale(LC_CTYPE,locale)) {
	printf("Can't set locale %s!\n",argv[2]);
	exit(1);
    }
    
    if (stat(argv[1],&st)) {
	printf("Specified file can't be stated!\n");
	exit(1);
    }
    
    if (!S_ISREG(st.st_mode)) {
	printf("Specified file isn't regular file!\n");
	exit(1);
    }
    
    text=(unsigned char*)malloc(st.st_size);
    if (!text) {
	printf("Can't allocate %lu bytes of memory!\n",st.st_size);
	exit(1);
    }
    
    f=fopen(argv[1],"r");
    if (!f) {
	printf("Failed to open specified file. Check permissions!\n");
	free(text);
	exit(1);
    }
    if (fread(text,1,st.st_size,f)!=st.st_size) {
	printf("Problem reading specified file!\n");
	free(text);
	fclose(f);
	exit(1);
    }
    fclose(f);
    
    for (i=0;i<st.st_size;i++) 
	text[i]=toupper(text[i]);
    
    f=fopen("UPPED.OUT","w");
    fwrite(text,1,st.st_size,f);
    fclose(f);
    free(text);
}


statgen/generate.c
#include <stdio.h>
#include <unistd.h>
#include <ctype.h>
#include <locale.h>
#include <math.h>
#include <iconv.h>
#include <langinfo.h>
#include <sys/types.h>
#include <sys/stat.h>

#define first_char 128
#define last_char 255

#define original_first_char 192
#define original_last_char 255

#define chars_number (last_char-first_char+1)
#define array_size (chars_number*chars_number)

struct array_pos {
    int ll;
    int uu;
    int lu;
    int ul;
};

struct pstat {
    unsigned long p;
    unsigned long s;
    unsigned long e;
};

iconv_t icnv=(iconv_t)-1;

int end_symbol(char ch) {
    if (ch=='\r'||ch=='\n'||ch==0||ch==' '||ch=='\t'||ch==','||ch=='.'||ch=='!'||ch=='?'||ch==';'||ch=='-'||ch==':'||ch=='"'||ch=='\''||ch==')') return 1;
    return 0;
}

int start_symbol(char ch) {
    if ((ch=='\t')||ch=='\r'||ch=='\n'||(ch==' ')||(ch=='(')||(ch=='"')||(ch=='\'')) return 1;
    return 0;
}    


unsigned char convert_char(unsigned char c) {
    char r;
    char *pr, *pc;
    size_t lr=1,lc=1;
    pr=&r;pc=&c;
    
    if (icnv == (iconv_t)-1) return c;
    if (iconv(icnv,&pc,&lc,&pr,&lr)<0) {
	printf("Error converting characters!\n");
	exit(1);
    }
    return r;
}

int get_array_pos(struct array_pos *pos, int a, int b) {
    int la,ua,lb,ub;
    if ((a<original_first_char)||(a>original_last_char)) return -1;
    if ((b<original_first_char)||(b>original_last_char)) return -1;
    
    la=tolower(a);
    ua=toupper(a);
    lb=tolower(b);
    ub=toupper(b);

    if ((la<original_first_char)||(la>original_last_char)) la=a;
    if ((lb<original_first_char)||(lb>original_last_char)) lb=b;
    if ((ua<original_first_char)||(ua>original_last_char)) ua=a;
    if ((ub<original_first_char)||(ub>original_last_char)) ub=b;
    
    la=convert_char(la);
    ua=convert_char(ua);
    lb=convert_char(lb);
    ub=convert_char(ub);
    
//    la=a;lb=b;ua=a;ub=b;
    
    pos->ll=(la-first_char)*chars_number+(lb-first_char);
    if (la!=ua) {
	pos->ul=(ua-first_char)*chars_number+(lb-first_char);
    } else {
	pos->ul=-1;
    } 
    if (lb!=ub) {
	pos->lu=(la-first_char)*chars_number+(ub-first_char);
    }
    else {
	pos->lu=-1;
    }
    if ((lb!=ub)&&(la!=ua)) {
	pos->uu=(ua-first_char)*chars_number+(ub-first_char);
    } else {
	pos->uu=-1;
    }
    return 0;
}


struct pstat *analyze(const unsigned char *text, unsigned long length) {
    struct pstat *a;
    unsigned long i;
    struct array_pos pos;
    
    a=(struct pstat*)malloc(array_size*sizeof(struct pstat));
    if (!a) return NULL;

    for (i=0;i<array_size;i++) {
	a[i].p=0;
	a[i].s=0;
	a[i].e=0;
    }
	
    for (i=1;i<length;i++) {
	if (get_array_pos(&pos,text[i-1],text[i])>=0) {
	    if (pos.ll>=0) {
		if ((i==1)||(start_symbol(text[i-2]))) a[pos.ll].s++;
		else if ((i+2==length)||(end_symbol(text[i+1]))) a[pos.ll].e++;
		else a[pos.ll].p++;
	    }
	    if (pos.ul>=0) {
		if ((i==1)||(start_symbol(text[i-2]))) a[pos.ul].s++;
		else if ((i+2==length)||(end_symbol(text[i+1]))) a[pos.ul].e++;
		else a[pos.ul].p++;
	    }
//	    if (pos.lu>=0) {
//		if ((i==1)||(start_symbol(text[i-2]))) a[pos.lu].s++;
//		else if ((i+2==length)||(end_symbol(text[i+1]))) a[pos.lu].e++;
//		else a[pos.lu].p++;
//	    }
	    if (pos.uu>=0) {
		if ((i==1)||(start_symbol(text[i-2]))) a[pos.uu].s++;
		else if ((i+2==length)||(end_symbol(text[i+1]))) a[pos.uu].e++;
		else a[pos.uu].p++;
	    }
	}
    }
    return a;
}


int print(struct pstat *a) {
    int i,j,k,n;
    
    for (i=first_char,k=0,n=0;i<=last_char;i++)
	for (j=first_char;j<=last_char;j++,k++) {
	    if ((a[k].p)||(a[k].s)||(a[k].e)) {
		if ((n)&&(n%8==0)) printf(",\n");
		else if (n) printf(", ");
		printf("{'%c','%c',%lf,%lf,%lf}",i,j,a[k].p?log10(a[k].p):-2,a[k].s?log10(a[k].s):-2,a[k].e?log10(a[k].e):-2);
		n++;
	    }
	}
    if ((n%8)!=1) printf("\n");
    return n;
}


unsigned long npow(unsigned long n) {
    unsigned long res=2;
    while (res<=n) res*=2;
    return res;
}

main(int argc, char *argv[]) {
    FILE *f;
    struct stat st;
    unsigned char *text;
    unsigned long len;
    struct pstat *a;
    int num;
    long i,sum;
    char locale[32];


    if (argc!=3) {
	printf("Usage: %s <file name> <encoding>\n",argv[0]);
	exit(0);
    }

    if (strlen(argv[2])>12) {
	printf("Invalid encoding(%s) specified!\n",argv[2]);
	exit(1);
    }
    
    if ((!strcasecmp(argv[2],"koi"))||(!strcasecmp(argv[2],"koi8"))||(!strcasecmp(argv[2],"koi-8"))||(!strcasecmp(argv[2],"koi8-r")))
    	sprintf(locale,"%s","KOI8-R");
    else if ((!strcasecmp(argv[2],"win"))||(!strcasecmp(argv[2],"cp1251"))||(!strcasecmp(argv[2],"cp-1251"))||(!strcasecmp(argv[2],"win1251"))||(!strcasecmp(argv[2],"win-1251")))
	sprintf(locale,"%s","CP1251");
    else if ((!strcasecmp(argv[2],"alt"))||(!strcasecmp(argv[2],"cp866"))||(!strcasecmp(argv[2],"cp-866"))||(!strcasecmp(argv[2],"ibm866"))||(!strcasecmp(argv[2],"ibm-866")))
	sprintf(locale,"%s","IBM866");
    else
	sprintf(locale,"%s",argv[2]);
    
    if (!setlocale(LC_CTYPE,"")) {
	printf("Can't set locale!\n");
	exit(1);
    }

    if (strcmp(locale,nl_langinfo(CODESET))) {
	if ((icnv=iconv_open(locale,nl_langinfo(CODESET)))<0) {
	    printf("Can't initialize iconv!\n");
	    exit(1);
	}
    }
    
    
    if (stat(argv[1],&st)) {
	printf("Specified file can't be stated!\n");
	iconv_close(icnv);
	exit(1);
    }
    
    if (!S_ISREG(st.st_mode)) {
	printf("Specified file isn't regular file!\n");
	iconv_close(icnv);
	exit(1);
    }
    
    text=(unsigned char*)malloc(st.st_size);
    if (!text) {
	printf("Can't allocate %lu bytes of memory!\n",st.st_size);
	iconv_close(icnv);
	exit(1);
    }
    
    f=fopen(argv[1],"r");
    if (!f) {
	printf("Failed to open specified file. Check permissions!\n");
	free(text);
	iconv_close(icnv);
	exit(1);
    }
    if (fread(text,1,st.st_size,f)!=st.st_size) {
	printf("Problem reading specified file!\n");
	free(text);
	fclose(f);
	iconv_close(icnv);
	exit(1);
    }
    fclose(f);
    
    a=analyze(text,st.st_size);
    if (a) {
	printf("static const lng_stat2 enc_%s[]={\n",argv[2]);
	num=print(a);
	printf("};\n\n");
	free(a);
	fprintf(stderr,"static unsigned int indexes2=%lu;\n",num);
	fprintf(stderr,"static unsigned int npow2=%lu;\n",npow(num));
    } else printf("Failed to allocate %lu bytes of memory!\n",array_size*sizeof(struct pstat));

    free(text);
    iconv_close(icnv);
}
